File: c/src/xorshift.c
#include "xorshift.h"

uint32_t xorshift32(struct xorshift32_state *state)
{
	/* Algorithm "xor" from p. 4 of Marsaglia, "Xorshift RNGs" */
	uint32_t x = state->a;
	x ^= x << 13;
	x ^= x >> 17;
	x ^= x << 5;
	return state->a = x;
}

float rand_float(struct xorshift32_state * state)
{
    uint32_t cur_value = xorshift32(state);
    uint32_t max_int = (uint32_t)-1;
    return (float)cur_value / (float)max_int;
};

File: java/Xorshift.java
package word_search;
public class Xorshift {
	Xorshift(int seed) {
		cur_state = seed;
	}

	int get_rand() {
		cur_state ^= cur_state << 13;
		cur_state ^= cur_state << 17;
		cur_state ^= cur_state >> 5;
		return cur_state;
	}

	float rand_float() {
		double cur_value = get_rand();
		double max_value = 0x00000000FFFFFFFFL;
		float out = (float)(cur_value / max_value);
		return out;
	}

	int cur_state;
}
